File: grokkingAlgorithms/javascript/dynamicProgramming/longestCommonSubsequence/longestCommonSubsequence.js
// Самая длинная общая подпоследовательность

/*
Описание: Допустим, вы открыли сайт dictionary.com. Пользователь вводит слово, а сайт возвращает определение.
Но если пользователь ввел несуществующее слово, нужно предположить, какое слово имелось ввиду.
Пользователь ищет определение fish, но он случайно ввел hish. Такого слова в словаре нет, но зато есть список похожих
слов: fish, vista.
*/

/**
 * Find the longest common subsequence
 * @param {string} firstWord First word
 * @param {string} secondWord Second word
 * @returns {number} The longest common subsequence
 */
const longestCommonSubsequence = (firstWord = "", secondWord = "") => {
    const matrix = JSON.parse(
        JSON.stringify(createMatrix(firstWord.length, secondWord.length))
    );
    if (matrix.length === 0 || matrix[0].length === 0) return 0;
    for (let i = 0; i < firstWord.length; i++) {
        for (let j = 0; j < secondWord.length; j++) {
            if (firstWord[i] === secondWord[j]) {
                matrix[i][j] = (i && j) > 0 ? matrix[i - 1][j - 1] + 1 : 1;
            } else {
                matrix[i][j] = Math.max(
                    i > 0 ? matrix[i - 1][j] : 0,
                    j > 0 ? matrix[i][j - 1] : 0
                );
            }
        }
    }
    return matrix[firstWord.length - 1][secondWord.length - 1];
};

longestCommonSubsequence("fish", "fosh"); // 3
longestCommonSubsequence("fort", "fosh"); // 2

File: grokkingAlgorithms/dynamicProgramming/longestCommonSubsequence/golang/longest_common_subsequence.go
package main

import "fmt"

// Самая длинная общая подстрока
func main() {
	fmt.Println(subsequence("fish", "fosh")) // 3
	fmt.Println(subsequence("fort", "fosh")) // 2
}

// createMatrix - функция для создания матрицы размера rows на cols
func createMatrix(rows, cols int) [][]int {
	cell := make([][]int, rows) // Создание массива, который будет хранить ссылки на строки

	for i := range cell {
		cell[i] = make([]int, cols)
	}

	return cell
}

// subsequence возвращает максимальную длину последовательности в строках.
func subsequence(a, b string) int {
	cell := createMatrix(len(a)+1, len(b)+1) // Создание матрицы размером len(a)+1 на len(b)+1

	for i := 1; i <= len(a); i++ { // Заполнение матрицы
		for j := 1; j <= len(b); j++ { // Проверка совпадения символов
			// Если символы совпадают, увеличиваем значение ячейки на 1
			if a[i-1] == b[j-1] {
				cell[i][j] = cell[i-1][j-1] + 1
				// Если символы не совпадают, выбираем максимальное значение из предыдущих ячеек
			} else {
				// Выбираем максимальное значение из ячеек выше и слева от текущей ячейки
				cell[i][j] = cell[i-1][j]
				// Если ячейка выше больше, чем ячейка слева, то присваиваем ячейке текущей ячейку выше
				if cell[i][j] < cell[i][j-1] {
					cell[i][j] = cell[i][j-1]
				}
			}
		}
	}

	// Возвращаем значение ячейки в правом нижнем углу, которая содержит максимальную длину последовательности
	return cell[len(a)][len(b)]
}
